IA-3762 useMetrics and Metrics TS improvements

msilva-broad opened the pull request:
- part of work for Export Analysis Modal Typescript and Unit Tests
- implemented useMetrics helper hook to remove boiler-plate code from event logging and improve Typescript ergonomics.
- removed now unnecessary @ts-expect-error escapes for Metrics.captureEvent calls in ts files.
- minor cleanup

msilva-broad committed:
IA-3762 useMetrics and Metrics TS improvements
- part of work for Export Analysis Modal Typescript and Unit Tests
- implemented useMetrics helper hook to remove boiler-plate code from event logging and improve Typescript ergonomics.
- removed now unnecessary @ts-expect-error escapes for Metrics.captureEvent calls in ts files.

msilva-broad commented on src/libs/ajax/Metrics.ts:
```diff
@@ -8,8 +8,8 @@ import { authStore, userStatus } from 'src/libs/state'
-export const Metrics = signal => ({
-  captureEvent: withErrorIgnoring(async (event, details = {}) => {
+export const Metrics = (signal?: AbortSignal) => {
+  const captureEventFn = async (event, details = {}) => {
```
breaking out this async function improves the Typescript flow out to the consumers

msilva-broad commented on src/components/common/IdContainer.ts:
```diff
@@ -7,7 +7,7 @@ type IdContainerProps = {
 
 /**
- * DEPRECATED - should switch to useUniqueIdFn pattern
+ * DEPRECATED - should switch to useUniqueId pattern (below)
```
unrelated... fixing a missed name update from previous PR

msilva-broad commented on src/components/file-browser/FileDetails.test.ts:
```diff
@@ -45,7 +45,9 @@ describe('FileDetails', () => {
-    const renderedValue = screen.getByText(field).parentElement!.lastElementChild!.textContent!
+    const rawRenderedValue = screen.getByText(field).parentElement!.lastElementChild!.textContent!
+    // normalize non-breaking spaces to just be normal space character for simpler assertion
+    const renderedValue = rawRenderedValue.replace(/\xa0/g, ' ').replace(/\u202f/g, ' ')
```
had to add this to tolerate non-breaking spaces used by date formatter on certain setups (mine).

nawatts commented on src/libs/ajax/metrics/useMetrics.ts:
```diff
@@ -0,0 +1,13 @@
+
+export type CaptureEventFn = (event: string, details?: {}) => void
+
+export const useMetricsEvent = (): CaptureEventFn => {
```
I don't understand the point of this hook.

> implemented useMetrics helper hook to remove boiler-plate code from event logging and improve Typescript ergonomics.

This seems like it adds boilerplate. Now I have to call the hook to get a function and then call that function instead of directly calling a function.

```js
const captureEvent = useMetricsEvent()
...
captureEvent(...)
```

```js
Ajax().Metrics.captureEvent(...)
```

msilva-broad replied:
The hook achieves a number of things:
- removes repetitive Ajax() constructor-like calls (efficiency win)
- is shorter in the main code by removing Ajax().Metrics prefix
- allows wrapping of void await 'fire and forget' to avoid TS cruft to awkwardly silence ts/lint errors/warnings (see comment below about 'why wrap this fn')
- far easier to mock, and removes unit-test boilerplate required to mock anything from Ajax() cluster.
- moves us to a more explicit dependency pattern instead of just Ajax().everythingUnderTheSun()

nawatts replied:
Could we accomplish all that and avoid the need for a hook / extra function call by exporting `captureEvent` from Metrics.js?

```js
import { captureEvent } from 'src/libs/ajax/Metrics'
...
captureEvent(...)
```

msilva-broad replied:
Looks like I edited the comment above just after you read it... 
But to your follow-up question... captureEvent is currently nested inside Ajax().Metrics sub-object's contract... I think your suggestion would break that pattern... which I don't think we want to just do willy-nilly.  But also, that would not bring over these benefits of the hook:
- wrapping void await to avoid ts/lint adherence cruft by consumers
- cleaner dependency and mocking pattern

nawatts replied:
The wrapper could be exported though. It seems like that would still improve mocking since you wouldn't have to mock the entire Ajax contract and would be more explicit about dependencies (import one function from Metrics instead of all of Ajax).

msilva-broad replied:
Ajax(signal).subContract pattern allows for Ajax(signal) instantiation-like flow-thru and other functional composition that I don't think we want to circumvent at this point.  While consumers of this specific function do not appear to make use of signal for cancelability,  a goal of this solution is to be conservative and additive for now.  Later discussion and proposals around code organization and patterns will come this year.

Also, (at the risk of sounding pedantic) choosing to fire-and-forget the promise feels like an application decision... These Ajax sub-contracts are pure in that they just expose promise-based contracts, leaving other concerns for elsewhere to determine.

I may be misinterpreting your suggestion though.

msilva-broad replied:
as we discussed on the call, another goal generally is to align dependency usage in our components into the pattern of useIntrestingData, useLogger, etc...  having a cleaner and narrower decoupling of each dependency... which is an iterative step towards key underpinnings of the larger cross-team code sharing initiative we are tackling this year.

nawatts replied:
Discussed in person. I was missing the bit that `useMetrics` separates components using metrics from the code in `libs/ajax` and `useMetrics` could be updated in the future to allow replacing the implementation of `captureEvent` via context, etc.

nawatts commented on src/libs/ajax/metrics/useMetrics.ts:
```diff
@@ -0,0 +1,13 @@
+
+export const useMetricsEvent = (): CaptureEventFn => {
+  const sendEvent = useMemo(() => Ajax().Metrics.captureEvent, [])
+  const captureEvent: CaptureEventFn = (event: string, details?: {}): void => {
```
Why does `sendEvent` need to be wrapped in `captureEvent` instead of returned directly?

msilva-broad replied:
this avoids some of the Typescript cruft by wrapping the fire-and-forget promise.  Without it Typescript and/or lint complains about the "no explicit void on await" on the consumer side, which is super awkward in some of the nested functional flows to accommodate.  So it's dealt with and wrapped here instead.

nawatts replied:
Oh, I see. Return void instead of Promise\<something\>. Perhaps worth a comment.

msilva-broad replied:
good call, ill add a comment

nawatts reviewed (commented):
Nice. Typing `captureEvent` gets rid of a frequent source of `ts-expect-error` comments.

I don't see the purpose of the `useMetricsEvent` hook though.

nawatts commented on src/libs/ajax/metrics/useMetrics.ts:
```diff
@@ -0,0 +1,13 @@
+  const captureEvent: CaptureEventFn = (event: string, details?: {}): void => {
+    void sendEvent(event, details)
+  }
+  return captureEvent
```
In the spirit of "delaying decisions", would it make sense to return an object `{ captureEvent }` instead of just `captureEvent` to allow adding additional methods/properties later?

msilva-broad replied:
i like it... since more of the metrics contract could come over into the hook at some point.

msilva-broad committed:
IAA-3762 useMetrics and Metrics TS improvements
- PR feedback
- added comments to hook implementation
- changed hook return to a contract object to accommodate easier addition of more metrics methods

s-rubenstein commented on src/libs/ajax/Metrics.ts:
```diff
@@ -31,14 +31,17 @@ export const Metrics = signal => ({
-  }),
+  }
+  return ({
+    captureEvent: (withErrorIgnoring(captureEventFn) as unknown as typeof captureEventFn),
```
Why do we need to do the unknown cast along the way? I've mostly seen that coercing between different generated objects of the same type (AoU Angular->React conversion)

msilva-broad replied:
this is always a last resort in my book... the hoF withErrorIgnoring had a bit too much lodash-fp type wiggliness to be cooperative without expanding the scope of the code changes into converting the related bits into TS as well.

n8landolt commented:
Looks like after my branch merged there's another //@ts-expect-error lying around.
